File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* dummy = new ListNode(0, head);

        ListNode* left = dummy;
        ListNode* right = dummy->next;

        for (int i = 0; i < n; i++) right = right->next;

        while (right) {
            right = right->next;
            left = left->next;
        }

        ListNode* point_none = left->next;
        left->next = left->next->next;
        delete point_none;

        ListNode* new_head = dummy->next;
        delete dummy;
        return new_head;
        
    }
};

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: ListNode | None, n: int) -> ListNode | None:

        dummy = ListNode(next = head)
        left_pointer = dummy
        right_pointer = dummy

        for _ in range(n+1):
            right_pointer = right_pointer.next
        
        while right_pointer:
            left_pointer = left_pointer.next
            right_pointer = right_pointer.next

        point_none = left_pointer.next
        left_pointer.next = left_pointer.next.next
        point_none.next = None


        return dummy.next
        